text/0000-app-namespacing-semantic-convention.md

```diff
@@ -0,0 +1,61 @@
+# Application namespacing semantic convention
+
+Adds a semantic convention for manual instrumentation that defines how to namespace custom attributes, starting with the `app` prefix.
+
+## Motivation
+
+Today, custom attributes from an app created through manual instrumentation have all kinds of names that are specific to the domain of that app. There are no conventions for how to organize this data, though.
+
+Data organization is critical to being able to effectively analyze observability data and avoid collisions in names. Without a well-defined convention for developers to follow, their own means of organization can differ from developer to developer, leading to a need to do things like coalesce field names using `.`/`-`/`_`/etc, not to mention having a real possibility to have name collisions. to separate clauses in an attribute name. Instead, a well-documented convention for how to organize data and in what format removes a "need to think" like code formatting conventions can do.
+
+## Explanation
+
+Attributes created through manual instrumentation should use namespacing for organizational purposes, using `app.` as a prefix for the name. Sub-namespaces should use a period (`.`) to separate sub-namespaces.
```

Comment:
One of the main purposes for namespaces is to avoid nameclashes. If everybody uses the same namespace, that kind of defeats the purpose. Also, something that is an "app" today might be a common library tomorrow.

Reply:
I don't think I understand this comment. You can get name clashes today without prefixing anything. But this is a way to better distinguish things that didn't come from an SDK or library or other component. How does this make name clashes worse?

Reply:
Not sure if this makes it worse, but I think it would be better to suggest that each app chooses their own root namespace for its attributes.

Reply:
I could see that being positive for folks to have flexibility. But I'm also leaning a bit into the dynamic of automated code formatters, which take certain decisions away and optionally give off-ramps from the opinionated default. What if instead of what's proposed today, the guidance is to choose your own root namespace, with `app` as a "blessed" example?

Reply:
+1 to reserving `app` as a default, protected-within-OTEL namespace, but encouraging users select their own prefix. This also implies that we ought to have a blacklist of protected attribute namespaces (e.g. `http.*`, `exception.*`), which can be inferred from [various](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions) [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions).

Reply:
Note: `app.*` was also suggested by the otel-client-side-telemetry SIG for semantic conventions of browser/mobile/desktop apps, not sure if this is still a thing (cc: @scheler, @martinkuba)

Beyond that +1:  I agree that it would be helpful to have namespaces that say "this is outside of the _offical_ semantic conventions". I assume that the potential is high, that without such a namespace a collision between future _official_ semantic conventions and what people are doing _custom_, is high: 

* My note above on dual use of `app` is actually an instance of a potential collision where the community wants to use something for a semantic convention while the same namespace is already used in the demo app (and others) for "custom attributes". 
* Think about any over-used word, turn it into a namespace and you might face the same problem in the future (`iot`, `mobile`,`mesh`,`runtime`,...)

I wrote "namespaces that say 'this is outside of the _offical_ semantic conventions'" above because I argue that there might be need for more than one:

* The one used by the end-user via manual instrumentations @cartermp is writing about in this OTEP (`app`, or some alternative, maybe `my`?)
* [There are already "protected" spaces for cloud vendors](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#cloud-provider-specific-attributes)
* There might be a similar use case for "vendors"/"distributions"

(Maybe that's overengineering that problem...)

> This also implies that we ought to have a blacklist of protected attribute namespaces (e.g. http.*, exception.*), which can be inferred from [various](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions) [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions).

💯 -- having some (soft/debug) checks that tell the user that they are using attributes out of scope (e.g. http.verb instead of http.method or db.order.id) would be a fantastic way to help people make good use of the semantic conventions.

Reply:
> Note: app.* was also suggested by the otel-client-side-telemetry SIG for semantic conventions of browser/mobile/desktop apps, not sure if this is still a thing

We tried introducing `app.name` in the past as a replacement to `service.name` in the context of client apps, since the term service doesn't make sense for client side apps. It was not accepted, since making `service.name` optional has other implications.